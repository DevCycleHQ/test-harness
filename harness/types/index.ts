export * from './capabilities'
export * from './sdks'
