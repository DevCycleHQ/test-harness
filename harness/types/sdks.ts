export const Sdks = {
  nodejs: "NodeJS",
  "of-nodejs": "OF-NodeJS",
  dotnet: "DotNet",
  go: "Go",
  ruby: "Ruby",
  php: "PHP",
  java: "Java",
  python: "Python",
};
