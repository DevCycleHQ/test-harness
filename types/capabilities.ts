export const Capabilities = {
    edgeDB: "EdgeDB",
    local: "LocalBucketing",
    cloud: "CloudBucketing",
    sse: "SSE"
}
