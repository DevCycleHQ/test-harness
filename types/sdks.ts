export enum Sdks {
    nodejs = "NodeJS",

}
