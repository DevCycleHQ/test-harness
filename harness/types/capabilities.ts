import { getSDKs } from './sdks'

export const Capabilities = {
    edgeDB: 'EdgeDB',
    cloud: 'CloudBucketing',
    sse: 'SSE',
    clientCustomData: 'ClientCustomData',
    multithreading: 'Multithreading',
    defaultReason: 'DefaultReason',
    etagReporting: 'EtagReporting',
    cloudProxy: 'CloudProxy',
    lastModifiedHeader: 'LastModifiedHeader',
    bootstrapping: 'Bootstrapping',
    sdkConfigEvent: 'SDKConfigEvent',
    clientUUID: 'ClientUUID',
    v2Config: 'V2Config',
    sdkPlatform: 'SDKPlatform',
    variablesFeatureId: 'VariableFeatureId',
    evalReason: 'EvalReason',
    allVariables: 'AllVariables',
    allFeatures: 'AllFeatures',
}

export const SDKPlatformMap = {
    'OF-NodeJS': 'nodejs-of',
}

//TODO: when tests are updated for OF-NodeJS eval capability, remove concept of allVariables/allFeatures as a capability
let sdkCapabilities: { [key: string]: string[] } = {
    NodeJS: [
        Capabilities.edgeDB,
        Capabilities.cloud,
        Capabilities.lastModifiedHeader,
        Capabilities.bootstrapping,
        Capabilities.sdkConfigEvent,
        Capabilities.clientUUID,
        Capabilities.v2Config,
        Capabilities.clientCustomData,
        Capabilities.variablesFeatureId,
        Capabilities.allVariables,
        Capabilities.allFeatures,
    ],
    'OF-NodeJS': [
        Capabilities.edgeDB,
        Capabilities.cloud,
        Capabilities.lastModifiedHeader,
        Capabilities.sdkConfigEvent,
        Capabilities.clientUUID,
        Capabilities.v2Config,
        Capabilities.sdkPlatform,
        Capabilities.variablesFeatureId,
    ],
    Python: [
        Capabilities.cloud,
        Capabilities.edgeDB,
        Capabilities.clientCustomData,
        Capabilities.v2Config,
        Capabilities.allVariables,
        Capabilities.allFeatures,
    ],
    DotNet: [
        Capabilities.cloud,
        Capabilities.edgeDB,
        Capabilities.clientCustomData,
        Capabilities.v2Config,
        Capabilities.allVariables,
        Capabilities.allFeatures,
        //Capabilities.sdkConfigEvent,
        //Capabilities.clientUUID,
    ],
    Java: [
        Capabilities.cloud,
        Capabilities.edgeDB,
        Capabilities.clientCustomData,
        Capabilities.v2Config,
        Capabilities.allVariables,
        Capabilities.allFeatures,
    ],
    Go: [
        Capabilities.cloud,
        Capabilities.edgeDB,
        Capabilities.clientCustomData,
        Capabilities.multithreading,
        Capabilities.defaultReason,
        Capabilities.etagReporting,
        Capabilities.lastModifiedHeader,
        Capabilities.sdkConfigEvent,
        Capabilities.clientUUID,
        Capabilities.v2Config,
        Capabilities.allVariables,
        Capabilities.allFeatures,
    ],
    Ruby: [
        Capabilities.clientCustomData,
        Capabilities.v2Config,
        Capabilities.variablesFeatureId,
        Capabilities.allVariables,
        Capabilities.allFeatures,
    ],
    PHP: [
        Capabilities.cloudProxy,
        Capabilities.allVariables,
        Capabilities.allFeatures,
    ],
}

export const getCapabilities = () => {
    const SDK_CAPABILITIES = process.env.SDK_CAPABILITIES

    if (!SDK_CAPABILITIES) {
        console.warn('No specified SDK Capabilities to test, running all tests')
        return Object.values(Capabilities)
    }

    try {
        const parsed = JSON.parse(SDK_CAPABILITIES)
        if (Array.isArray(parsed)) {
            return parsed.map((capability) => Capabilities[capability])
        } else if (typeof parsed === 'object' && parsed !== null) {
            // Return as-is for per-SDK logic
            return Object.fromEntries(
                Object.entries(parsed).map(([sdk, caps]) => [
                    sdk,
                    Array.isArray(caps)
                        ? caps.map((cap) => Capabilities[cap]).filter(Boolean)
                        : [],
                ]),
            )
        }
    } catch (e) {
        if (SDK_CAPABILITIES && Capabilities[SDK_CAPABILITIES]) {
            return [Capabilities[SDK_CAPABILITIES]]
        } else if (SDK_CAPABILITIES) {
            return SDK_CAPABILITIES.split(',')
                .map((capability) => Capabilities[capability])
                .filter((capability) => capability !== undefined)
        }
    }
    return Object.values(Capabilities)
}

if (process.env.SDK_CAPABILITIES && process.env.SDKS_TO_TEST) {
    const sdks: string[] = getSDKs()
    const capabilities = getCapabilities()
    if (Array.isArray(capabilities)) {
        sdkCapabilities = sdks.reduce((acc, sdk) => {
            acc[sdk] = capabilities
            return acc
        }, {})
    } else if (typeof capabilities === 'object' && capabilities !== null) {
        sdkCapabilities = sdks.reduce((acc, sdk) => {
            acc[sdk] = capabilities[sdk] || sdkCapabilities[sdk] || []
            return acc
        }, {})
    }
}

export const SDKCapabilities = sdkCapabilities
