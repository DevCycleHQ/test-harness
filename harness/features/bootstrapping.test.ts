import {
    describeCapability,
    getSDKScope,
    hasCapability,
    LocalTestClient,
} from '../helpers'
import { Capabilities } from '../types'
import immutable from 'object-path-immutable'

describe('Bootstrapping Tests', () => {
    const { sdkName, scope } = getSDKScope()

    describeCapability(sdkName, Capabilities.bootstrapping)(sdkName, () => {
        let testClient: LocalTestClient

        describe('not enabled client', () => {
            beforeEach(async () => {
                testClient = new LocalTestClient(sdkName, scope)

                if (hasCapability(sdkName, Capabilities.sdkConfigEvent)) {
                    scope
                        .post(`/client/${testClient.clientId}/v1/events/batch`)
                        .reply(201, {
                            message: 'Successfully received events.',
                        })
                }

                await testClient.createClient(true, {
                    configPollingIntervalMS: 100000,
                    eventFlushIntervalMS: 500,
                })
            })

            it('throws an error when trying to get bootstrap config if not enabled', async () => {
                await testClient.callGetClientBootstrapConfig(
                    {
                        user_id: 'test',
                    },
                    '',
                    true,
                )
            })
        })

        describe('initialized client', () => {
            beforeEach(async () => {
                testClient = new LocalTestClient(sdkName, scope)

                if (hasCapability(sdkName, Capabilities.sdkConfigEvent)) {
                    scope
                        .post(`/client/${testClient.clientId}/v1/events/batch`)
                        .reply(201, {
                            message: 'Successfully received events.',
                        })
                }

                // create a different config clientside so we can make sure the bootstrapping method is using the right one
                const clientsideConfig = {
                    ...immutable.set(
                        testClient.getValidConfig(),
                        'features.0.variations.0.variables.1.value',
                        'new string',
                    ),
                    clientSDKKey: 'client-key',
                }

                scope
                    .get(
                        `/client/${testClient.clientId}/config/v1/server/bootstrap/${testClient.sdkKey}.json`,
                    )
                    .reply(200, clientsideConfig)

                await testClient.createClient(true, {
                    configPollingIntervalMS: 100000,
                    eventFlushIntervalMS: 500,
                    enableClientBootstrapping: true,
                })
            })

            it('should return a client bootstrapped config', async () => {
                const configResponse =
                    await testClient.callGetClientBootstrapConfig(
                        {
                            user_id: 'test',
                            customData: {
                                'should-bucket': true,
                            },
                        },
                        '',
                    )

                const configResult = await configResponse.json()

                expect(configResult.data.variables['string-var'].value).toEqual(
                    'new string',
                )
                expect(configResult.data.clientSDKKey).toEqual('client-key')
            })
        })
    })
})
