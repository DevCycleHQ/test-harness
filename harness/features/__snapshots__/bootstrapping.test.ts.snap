// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bootstrapping Tests NodeJS initialized client should return a client bootstrapped config 1`] = `
{
  "data": {
    "environment": {
      "_id": "638680c459f1b81cc9e6c559",
      "key": "development",
    },
    "featureVariationMap": {
      "638680d6fcb67b96878d90e6": "638680d6fcb67b96878d90ec",
      "6386813a59f1b81cc9e6c68d": "6386813a59f1b81cc9e6c693",
    },
    "features": {
      "schedule-feature": {
        "_id": "6386813a59f1b81cc9e6c68d",
        "_variation": "6386813a59f1b81cc9e6c693",
        "key": "schedule-feature",
        "type": "release",
        "variationKey": "variation-on",
        "variationName": "Variation On",
      },
      "test-harness": {
        "_id": "638680d6fcb67b96878d90e6",
        "_variation": "638680d6fcb67b96878d90ec",
        "key": "test-harness",
        "type": "release",
        "variationKey": "variation-on",
        "variationName": "Variation On",
      },
    },
    "project": {
      "_id": "638680c459f1b81cc9e6c557",
      "a0_organization": "org_fakeorg",
      "key": "test-harness-data",
      "settings": {
        "edgeDB": {
          "enabled": false,
        },
        "optIn": {
          "colors": {
            "primary": "#000000",
            "secondary": "#000000",
          },
          "enabled": false,
        },
      },
    },
    "variableVariationMap": {
      "bool-var": {
        "_feature": "638680d6fcb67b96878d90e6",
        "_variation": "638680d6fcb67b96878d90ec",
      },
      "json-var": {
        "_feature": "638680d6fcb67b96878d90e6",
        "_variation": "638680d6fcb67b96878d90ec",
      },
      "number-var": {
        "_feature": "638680d6fcb67b96878d90e6",
        "_variation": "638680d6fcb67b96878d90ec",
      },
      "schedule-feature": {
        "_feature": "6386813a59f1b81cc9e6c68d",
        "_variation": "6386813a59f1b81cc9e6c693",
      },
      "string-var": {
        "_feature": "638680d6fcb67b96878d90e6",
        "_variation": "638680d6fcb67b96878d90ec",
      },
      "unicode-var": {
        "_feature": "638680d6fcb67b96878d90e6",
        "_variation": "638680d6fcb67b96878d90ec",
      },
    },
    "variables": {
      "bool-var": {
        "_id": "638681f059f1b81cc9e6c7fa",
        "key": "bool-var",
        "type": "Boolean",
        "value": true,
      },
      "json-var": {
        "_id": "638681f059f1b81cc9e6c7fd",
        "key": "json-var",
        "type": "JSON",
        "value": {
          "facts": true,
        },
      },
      "number-var": {
        "_id": "638681f059f1b81cc9e6c7fc",
        "key": "number-var",
        "type": "Number",
        "value": 1,
      },
      "schedule-feature": {
        "_id": "6386813a59f1b81cc9e6c68f",
        "key": "schedule-feature",
        "type": "Boolean",
        "value": true,
      },
      "string-var": {
        "_id": "638681f059f1b81cc9e6c7fb",
        "key": "string-var",
        "type": "String",
        "value": "new string",
      },
      "unicode-var": {
        "_id": "638681f059f1b81cc9e6c7fe",
        "key": "unicode-var",
        "type": "String",
        "value": "‚Üë‚Üë‚Üì‚Üì‚Üê‚Üí‚Üê‚ÜíBA ü§ñ",
      },
    },
  },
  "entityType": "Object",
  "logs": [],
}
`;
